Add colouring to wrong words

ailin-nemui opened the pull request:
Highlight:

wrong words can be highlighted in the input prompt:

![spellcheck](https://user-images.githubusercontent.com/5665186/48140655-8431b280-e2a9-11e8-8778-f20571eff325.png)

jwilk commented on spellcheck.pl:
```diff
@@ -16,9 +16,11 @@
-$VERSION = '0.8.1';
+$VERSION = '0.8.2';
 %IRSSI = (
     authors     => 'Jakub Wilk, Jakub Jankowski, Gabriel Pettier',
```
Please consider adding yourself to the list of authors (and to the list of copyright holders earlier in the file).

jwilk commented on spellcheck.pl:
```diff
@@ -60,6 +62,9 @@ sub spellcheck_check_word
 
     return if $word =~ m{^/}; # looks like a path
+    if (Irssi::settings_get_bool('expand_escapes')) {
+	$word =~ s/\\(['"\\])/$1/g;
```
I don't understand why is this needed.
Please use spaces, not tabs.

ailin-nemui replied:
it's another bug I fixed, if this setting is enabled we should replace \\'  with  '  when spell checking words such as "it's"

jwilk commented on spellcheck.pl:
```diff
@@ -135,30 +140,39 @@ sub spellcheck_key_pressed
     # hide correction window when message is sent
     if (chr($key) =~ /\A[\r\n]\z/ && $correction_window) {
         $correction_window->command("window hide $window_name");
+        if (Irssi->can('gui_input_clear_extents')) {
```
This API doesn't seem to be present in the latest Irssi release (1.1.1). Is that correct?
In any case, minimum version requirements should be documented in the commit messages, and eventually also in README.

ailin-nemui replied:
try irssi from git / 1.2

jwilk commented on spellcheck.pl:
```diff
@@ -135,30 +140,39 @@ sub spellcheck_key_pressed
+
+    # get current inputline
+    my $inputline = Irssi::parse_special('$L');
+    if (Irssi::settings_get_str('term_charset') =~ /utf-?8/i) {
```
The regexp should be anchored (with `\A` and `\z`)…

jwilk commented on spellcheck.pl:
```diff
@@ -135,30 +140,39 @@ sub spellcheck_key_pressed
+    # get current inputline
+    my $inputline = Irssi::parse_special('$L');
+    if (Irssi::settings_get_str('term_charset') =~ /utf-?8/i) {
+        Encode::_utf8_on($inputline);
```
…but maybe it would be better to just decode the input line using `term_charset` encoding? Then you wouldn't need any ifs or regexps.

ailin-nemui replied:
I don't think that's economic

jwilk commented on spellcheck.pl:
```diff
@@ -135,30 +140,39 @@ sub spellcheck_key_pressed
+        Encode::_utf8_on($inputline);
+    }
+    if (Irssi->can('gui_input_set_extent')) {
+	Irssi::gui_input_set_extent(length $inputline, '%n');
```
tabs → spaces

jwilk commented on spellcheck.pl:
```diff
@@ -135,30 +140,39 @@ sub spellcheck_key_pressed
-    my $cmdchars = Irssi::settings_get_str('cmdchars');
-    my $re = qr/^[$cmdchars]/;
-    return if ($inputline =~ $re);
+    return unless -1 == index(Irssi::settings_get_str('cmdchars'), substr($inputline, 0, 1));
```
I guess the original code will break if `cmdchars` contains `\` or `]`.
Is that the only bug you're fixing here?
Anyway, this seems unrelated to coloring words on input line, so it should be in a separate commit.

ailin-nemui replied:
you're right I fixed a bug here

jwilk replied:
I commited a different fix for this: 62b1fdb561ff5387c2374170ff7745b369eadb2b
Thanks for the bug report.

jwilk commented on spellcheck.pl:
```diff
@@ -135,30 +140,39 @@ sub spellcheck_key_pressed
 
     # get last bit from the inputline
-    my ($word) = $inputline =~ /\s*(\S+)$/;
+    my ($word) = $inputline =~ /\s*(\S+)[.\s]*$/;
```
Why?

ailin-nemui replied:
the input line will now end with what we typed, so we should ignore/allow space and .

jwilk replied:
Ignoring the dot doesn't work, because it will be consumed by `\S+`. But meh, we can fix it later.

jwilk commented on spellcheck.pl:
```diff
@@ -168,10 +182,18 @@ sub spellcheck_key_pressed
-        $win->command("window show $window_name");
-        $correction_window->command('window stick off');
+        $win->command("^window show $window_name");
+        $correction_window->command('^window stick off');
```
What does leading `^` do?

ailin-nemui replied:
hide messages like window is not sticky/window is already visible

jwilk commented on spellcheck.pl:
```diff
@@ -168,10 +182,18 @@ sub spellcheck_key_pressed
-        $correction_window->command('window stick off');
+        $win->command("^window show $window_name");
+        $correction_window->command('^window stick off');
+        $win->set_active;
```
Why?

ailin-nemui replied:
another bug fix, when we run /win show, the input focus will change to the "spell" window. but we do not want this, so set the focus back to the original window

jwilk commented on spellcheck.pl:
```diff
@@ -239,13 +261,17 @@ sub spellcheck_add_word
-Irssi::signal_add_first('gui key pressed', 'spellcheck_key_pressed');
+Irssi::signal_add_last('key word_completion', sub{spellcheck_key_pressed(ord '.')});
+Irssi::signal_add_last('key word_completion_backward', sub{spellcheck_key_pressed(ord '.')});
+Irssi::signal_add_last('gui key pressed', 'spellcheck_key_pressed');
```
What do the changes to `signal_add_…` do?

ailin-nemui replied:
the highlight should be updated when you use tab (word_completion), and it should also encompass the last typed letter (thus last instead of first)

jwilk reviewed (changes requested):
I haven't tested this yet, but the screenshot looks awesome.

The author email for this commit is `ailin@z30a.localdomain`, which isn't associated with your GitHub account and doesn't look correct all.

ailin-nemui committed:
Add colouring to wrong words

words which are not recognised by the spell checker can be optionally
highlighted in the input prompt.

Two new settings have been added:
- `spellcheck_word_input_color', which selects the colour to be used on
  the input line; and
- `spellcheck_print_suggestions', which can be used to turn off printing
  the suggestions in case highlighting alone is deemed sufficient.

This feature can be used on Irssi 1.2.0 and later.

ailin-nemui commented:
I attempted to implement your suggestions.

jwilk commented on spellcheck.pl:
```diff
@@ -60,6 +60,9 @@ sub spellcheck_check_word
 
     return if $word =~ m{^/}; # looks like a path
+    if (Irssi::settings_get_bool('expand_escapes')) {
+        $word =~ s/\\(['"\\])/$1/g;
```
I don't think that's right.
My Irssi (1.0.7) doesn't expand `\'` to `'` even when `expand_escapes` is enabled.
I've filed separate bug (https://github.com/jwilk/irssi-spellcheck/issues/2) for proper support for `expand_escapes`.

ailin-nemui replied:
I see, we're both right. This was due to the fix that landed in Irssi 1.1.0 https://github.com/irssi/irssi/pull/709 but I think it won't hurt on earlier versions

ailin-nemui replied:
I dropped this patch as it doesn't make sense and the Irssi feature has a bug

jwilk commented:
Can I have your full name in git commits metadata and in the `%IRSSI` hash? (I accept "no" as an anwer.)

jwilk commented on spellcheck.pl:
```diff
@@ -239,13 +260,17 @@ sub spellcheck_add_word
 Irssi::settings_add_str( 'spellcheck', 'spellcheck_word_color', '%R');
 Irssi::settings_add_str( 'spellcheck', 'spellcheck_window_name', '');
 Irssi::settings_add_str( 'spellcheck', 'spellcheck_window_height', 10);
+Irssi::settings_add_str( 'spellcheck', 'spellcheck_word_input_color', '%U');
```
Any particular reason the default is underline, rather than red? I think on the screenshot red looked pretty good.

ailin-nemui replied:
just what I liked

ailin-nemui commented:
I have several scripts published with my nick as author so I'd leave it

jwilk commented on spellcheck.pl:
```diff
@@ -170,8 +170,9 @@ sub spellcheck_key_pressed
-        $win->command("window show $window_name");
-        $correction_window->command('window stick off');
+        $win->command("^window show $window_name");
+        $correction_window->command('^window stick off');
```
I fixed another similar issue and commited it as 7c06d8c507c1824896d13cc2a2914efde1c7fdb9.

jwilk commented on spellcheck.pl:
```diff
@@ -170,8 +170,9 @@ sub spellcheck_key_pressed
-        $correction_window->command('window stick off');
+        $win->command("^window show $window_name");
+        $correction_window->command('^window stick off');
+        $win->set_active;
```
Cherry-picked as 4efbb19f0a75b306eedb0925475fed552bda936c.

jwilk commented on spellcheck.pl:
```diff
@@ -16,11 +16,13 @@
 use Text::Aspell;
 
-$VERSION = '0.8.1';
+$VERSION = '0.8.2';
```
0.8.1 hasn't been released yet, so there was no need to bump this.

jwilk commented on spellcheck.pl:
```diff
@@ -135,20 +137,27 @@ sub spellcheck_key_pressed
+        Encode::_utf8_on($inputline);
+    }
+    if (Irssi->can('gui_input_set_extent')) {
+        Irssi::gui_input_set_extent(length $inputline, '%n');
```
What was this supposed to do?
It seems to make sure that when you correct an already-colored word, the newly added letters are not colored. But this only works at the end of the input line.

ailin-nemui replied:
good point. you need to ensure that the line ends with %n always, otherwise if you delete the last character the underline will go to the end of the input prompt. but in fact the same issue will happen if you delete the last character of an underlined word :thinking:

jwilk commented:
> I have several scripts published with my nick as author

Fair enough, but please be consistent. f94e80fb14d11342e71307d11120dcf4a385c039 has a different author than ca448ca45cb7d2c6dd402f16f2d67fe92a9074d8.

ailin-nemui committed:
add myself to list of authors

jwilk commented:
Merged as eb33fe308686cbd0e062c144d1719a43115be714 and 281d7ea1cb3e14fb7bef409ee95d91fdde53ea9a.

There are still some rough edges, but we can fix them later.

Thanks for implementing this excellent feature and sorry for taking so long to merge it!
